### PR TITLE
[Megatron dataset] Support loading megatron dataset

### PR DESCRIPTION
**Pull request by KB-Ding**
### PR types

New features
### PR changes

APIs
### Description

1. Support loading megatron dataset for [llm/llama](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files), [llm/gpt3](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files) and [model_zoo/gpt/run_pretrain_trainer.py.](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files)
2. Add [paddlenlp/data/indexed_dataset.py](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files) for megatron.
3. Add [paddlenlp/data/causal_dataset.py](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files) for llm.
4. Add [paddlenlp/data/blendable_dataset.py](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files) for multi-dataset.
5. Add [model_zoo/gpt/dataset.py](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files) so that [model_zoo/gpt/run_pretrain_trainer.py](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files) can load megatron dataset.
6. Add [model_zoo/gpt/legacy_dataset.py](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files) for [model_zoo/gpt/run_pretrain_static.py](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files) and [model_zoo/gpt/run_pretrain.py](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files).
7. Remove duplicate content in [ernie-1.0/data_tools](https://github.com/PaddlePaddle/PaddleNLP/pull/6489/files).
8. Support custom data_cache path.

### CODE REVIEW

File: llm/gpt-3/run_pretrain.py [hunk 1]
```diff
@@ -99,6 +99,12 @@ class DataArguments:
         metadata={"help": "Use share folder for data dir and output dir on multi machine."},
     )
 
+    data_impl: str = field(default="mmap", metadata={"help": "Data implementation."})
```
Comment (gongel) [inline llm/gpt-3/run_pretrain.py]: help描述可以更清楚, 可以增加choice限定范围

File: llm/gpt-3/run_pretrain.py [hunk 2]
```diff
@@ -99,6 +99,12 @@ class DataArguments:
...
+    skip_warmup: bool = field(
+        default=True,
+        metadata={"help": "Skip warmup or not."},
```
Comment (gongel) [inline llm/gpt-3/run_pretrain.py]: 这个help同上

File: llm/gpt-3/run_pretrain.py [hunk 3]
```diff
@@ -138,7 +144,7 @@ def create_pretrained_dataset(
     tokenizer,
 ):
 
-    train_valid_test_num_samples = [
+    train_val_test_num_samples = [
```
Comment (gongel) [inline llm/gpt-3/run_pretrain.py:152]: 换名字的目的是？

File: llm/gpt-3/run_pretrain.py [hunk 4]
```diff
@@ -150,72 +156,55 @@ def create_pretrained_dataset(
...
 
     def _collate_data(data, stack_fn=Stack()):
-        num_fields = len(data[0])
-        out = [None] * num_fields
-        # 0:input_ids, 1:loss_mask, 2:attention_mask, 3:position_ids, 4:labels
-        for i in (0, 1, 2, 3, 4):
-            out[i] = stack_fn([x[i] for x in data])
+        tokens_ = stack_fn(x["text"] for x in data)
+        # Unpack.
+        tokens_ = paddle.to_tensor(tokens_, dtype="int64")
```
Comment (gongel) [inline llm/gpt-3/run_pretrain.py]: 这个去掉吧，Stack()出来后就是tensor。不同paddle版本可能会出错

File: llm/gpt-3/run_pretrain.py [hunk 5]
```diff
@@ -150,72 +156,55 @@ def create_pretrained_dataset(
...
+
+        # Attention mask.
+        attention_mask = paddle.ones(tokens.shape, dtype=paddle.int64)
 
         return {
-            "input_ids": out[0],
-            "attention_mask": out[2],
-            "labels": out[4],
+            "input_ids": tokens,
+            # "token_type_ids": out[1],
```
Comment (gongel) [inline llm/gpt-3/run_pretrain.py]: 不要的可以删掉

**Review by gongel — COMMENTED** [review]
LLaMA的comment同GPT-3，冲突需要解决下。

File: llm/llama/run_pretrain.py [hunk 6]
```diff
@@ -110,6 +109,13 @@ class DataArguments:
         metadata={"help": "Use share folder for data dir and output dir on multi machine."},
     )
 
+    data_impl: str = field(default="mmap", metadata={"help": "The format of the preprocessed data."})
```
Comment (gongel) [inline llm/llama/run_pretrain.py:113]: 这里加个 choice，限定范围

File: llm/llama/run_pretrain.py [hunk 7]
```diff
@@ -202,74 +208,56 @@ def create_pretrained_dataset(
...
-        for i in (0, 1, 2, 3, 4):
-            out[i] = stack_fn([x[i] for x in data])
+        tokens_ = stack_fn(x["text"] for x in data)
+        # Unpack.
+        # tokens_ = paddle.to_tensor(tokens_, dtype="int64")
+        labels = tokens_[:, 1:]
+        tokens = tokens_[:, :-1]
+
+        # Loss mask.
+        loss_mask = paddle.ones(tokens.shape, dtype=paddle.float32)
```
Comment (gongel) [inline llm/llama/run_pretrain.py]: 这里用不到的话，可以删掉了，`eos_token_id`也不用引入了

File: llm/gpt-3/run_pretrain.py (hunk 3 above)
Comment (KB-Ding) [reply llm/gpt-3/run_pretrain.py:152]: 已修改回来

File: llm/gpt-3/run_pretrain.py (hunk 4 above)
Comment (KB-Ding) [reply llm/gpt-3/run_pretrain.py]: Done

File: llm/gpt-3/run_pretrain.py (hunk 5 above)
Comment (KB-Ding) [reply llm/gpt-3/run_pretrain.py]: Done

File: llm/llama/run_pretrain.py (hunk 7 above)
Comment (KB-Ding) [reply llm/llama/run_pretrain.py]: Done

File: llm/gpt-3/run_pretrain.py (hunk 1 above)
Comment (KB-Ding) [reply llm/gpt-3/run_pretrain.py]: 目前限定比较麻烦，先走默认的mmap。

**Review by gongel — DISMISSED** [review]

**Review by wawltor — APPROVED** [review]
LGTM

**Pull request merged**